### PR TITLE
fix: replace p-queue with less restrictive queue

### PR DESCRIPTION
**Pull request by achingbrain**
Adds a `Queue` class to `@libp2p/utils` modelled on p-queue with a few key differences:

1. The queue is externally accessible so we can modify it before jobs run
2. It can be turned into an async generator
3. Jobs remain in the queue while they are executing for better introspection
4. It integrates with libp2p metrics, if desired

The dial queue has been replaced with this new queue class, this means we don't need to maintain a separate internal queue for pending dials since the dial queue itself is accessible.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/utils/src/queue/index.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+  async onIdle (): Promise<void> {
+    // Instantly resolve if none pending and if nothing else is queued
+    if (this.pending === 0 && this.size === 0) {
+      return
+    }
+
+    await this.onEvent('idle')
+  }
+
+  private async onEvent (event: keyof QueueEvents<JobReturnType>, filter?: () => boolean): Promise<void> {
```
Comment (wemeetagain) [inline packages/utils/src/queue/index.ts:276]: Does this need to be abortable?
I guess not if clear/abort will allow all pending promises (from public methods, onEmpty, onSizeLessThan, onIdle, toGenerator) to resolve.

Comment (achingbrain) [reply packages/utils/src/queue/index.ts:276]: It's not abortable in p-queue, though it makes sense

Comment (achingbrain) [reply packages/utils/src/queue/index.ts]: They are abortable now

File: packages/utils/package.json [hunk 2]
```diff
@@ -76,9 +76,13 @@
...
-    "./peer-job-queue": {
-      "types": "./dist/src/peer-job-queue.d.ts",
-      "import": "./dist/src/peer-job-queue.js"
+    "./queue": {
+      "types": "./dist/src/queue/index.d.ts",
+      "import": "./dist/src/queue/index.js"
+    },
+    "./peer-queue": {
+      "types": "./dist/src/peer-queue.d.ts",
+      "import": "./dist/src/peer-queue.js"
```
Comment (akim-bow) [inline packages/utils/package.json:85]: @achingbrain This PR introduces BREAKING CHANGE and actually broke many libp2p projects. Why not to bump major version if you break package API? Libp2p packages use caret ranges and automatically will try to update this package when installed.

**Pull request merged**